# Исследование надежности заемщиков


Во второй части проекта вы выполните шаги 3 и 4. Их вручную проверит ревьюер.
Чтобы вам не пришлось писать код заново для шагов 1 и 2, мы добавили авторские решения в ячейки с кодом. 



## Откройте таблицу и изучите общую информацию о данных

**Задание 1. Импортируйте библиотеку pandas. Считайте данные из csv-файла в датафрейм и сохраните в переменную `data`. Путь к файлу:**

`/datasets/data.csv`

In [1]:
import pandas as pd

try:
    data = pd.read_csv('/datasets/data.csv')
except:
    data = pd.read_csv('https://code.s3.yandex.net/datasets/data.csv')

**Задание 2. Выведите первые 20 строчек датафрейма `data` на экран.**

In [2]:
data.head(20)

children  days_employed  dob_years            education  education_id  \
0          1   -8437.673028         42               высшее             0   
1          1   -4024.803754         36              среднее             1   
2          0   -5623.422610         33              Среднее             1   
3          3   -4124.747207         32              среднее             1   
4          0  340266.072047         53              среднее             1   
5          0    -926.185831         27               высшее             0   
6          0   -2879.202052         43               высшее             0   
7          0    -152.779569         50              СРЕДНЕЕ             1   
8          2   -6929.865299         35               ВЫСШЕЕ             0   
9          0   -2188.756445         41              среднее             1   
10         2   -4171.483647         36               высшее             0   
11         0    -792.701887         40              среднее             1   
12         0            NaN         65              среднее             1   
13         0   -1846.641941         54  неоконченное высшее             2   
14         0   -1844.956182         56               высшее             0   
15         1    -972.364419         26              среднее             1   
16         0   -1719.934226         35              среднее             1   
17         0   -2369.999720         33               высшее             0   
18         0  400281.136913         53              среднее             1   
19         0  -10038.818549         48              СРЕДНЕЕ             1   

       family_status  family_status_id gender income_type  debt  \
0    женат / замужем                 0      F   сотрудник     0   
1    женат / замужем                 0      F   сотрудник     0   
2    женат / замужем                 0      M   сотрудник     0   
3    женат / замужем                 0      M   сотрудник     0   
4   гражданский брак                 1      F   пенсионер     0   
5   гражданский брак                 1      M   компаньон     0   
6    женат / замужем                 0      F   компаньон     0   
7    женат / замужем                 0      M   сотрудник     0   
8   гражданский брак                 1      F   сотрудник     0   
9    женат / замужем                 0      M   сотрудник     0   
10   женат / замужем                 0      M   компаньон     0   
11   женат / замужем                 0      F   сотрудник     0   
12  гражданский брак                 1      M   пенсионер     0   
13   женат / замужем                 0      F   сотрудник     0   
14  гражданский брак                 1      F   компаньон     1   
15   женат / замужем                 0      F   сотрудник     0   
16   женат / замужем                 0      F   сотрудник     0   
17  гражданский брак                 1      M   сотрудник     0   
18    вдовец / вдова                 2      F   пенсионер     0   
19         в разводе                 3      F   сотрудник     0   

     total_income                                 purpose  
0   253875.639453                           покупка жилья  
1   112080.014102                 приобретение автомобиля  
2   145885.952297                           покупка жилья  
3   267628.550329              дополнительное образование  
4   158616.077870                         сыграть свадьбу  
5   255763.565419                           покупка жилья  
6   240525.971920                       операции с жильем  
7   135823.934197                             образование  
8    95856.832424                   на проведение свадьбы  
9   144425.938277                 покупка жилья для семьи  
10  113943.491460                    покупка недвижимости  
11   77069.234271       покупка коммерческой недвижимости  
12            NaN                         сыграть свадьбу  
13  130458.228857                 приобретение автомобиля  
14  165127.911772              покупка жилой недвижимости  
15  116820.904450 

**Задание 3. Выведите основную информацию о датафрейме с помощью метода `info()`.**

In [3]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


## Предобработка данных

### Удаление пропусков

**Задание 4. Выведите количество пропущенных значений для каждого столбца. Используйте комбинацию двух методов.**

In [4]:
data.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

**Задание 5. В двух столбцах есть пропущенные значения. Один из них — `days_employed`. Пропуски в этом столбце вы обработаете на следующем этапе. Другой столбец с пропущенными значениями — `total_income` — хранит данные о доходах. На сумму дохода сильнее всего влияет тип занятости, поэтому заполнить пропуски в этом столбце нужно медианным значением по каждому типу из столбца `income_type`. Например, у человека с типом занятости `сотрудник` пропуск в столбце `total_income` должен быть заполнен медианным доходом среди всех записей с тем же типом.**

In [5]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['total_income'].isna()), 'total_income'] = \
    data.loc[(data['income_type'] == t), 'total_income'].median()

### Обработка аномальных значений

**Задание 6. В данных могут встречаться артефакты (аномалии) — значения, которые не отражают действительность и появились по какой-то ошибке. таким артефактом будет отрицательное количество дней трудового стажа в столбце `days_employed`. Для реальных данных это нормально. Обработайте значения в этом столбце: замените все отрицательные значения положительными с помощью метода `abs()`.**

In [6]:
data['days_employed'] = data['days_employed'].abs()

**Задание 7. Для каждого типа занятости выведите медианное значение трудового стажа `days_employed` в днях.**

In [7]:
data.groupby('income_type')['days_employed'].agg('median')

income_type
безработный        366413.652744
в декрете            3296.759962
госслужащий          2689.368353
компаньон            1547.382223
пенсионер          365213.306266
предприниматель       520.848083
сотрудник            1574.202821
студент               578.751554
Name: days_employed, dtype: float64

У двух типов (безработные и пенсионеры) получатся аномально большие значения. Исправить такие значения сложно, поэтому оставьте их как есть. Тем более этот столбец не понадобится вам для исследования.

**Задание 8. Выведите перечень уникальных значений столбца `children`.**

In [8]:
data['children'].unique()

array([ 1,  0,  3,  2, -1,  4, 20,  5])

**Задание 9. В столбце `children` есть два аномальных значения. Удалите строки, в которых встречаются такие аномальные значения из датафрейма `data`.**

In [9]:
data = data[(data['children'] != -1) & (data['children'] != 20)]

**Задание 10. Ещё раз выведите перечень уникальных значений столбца `children`, чтобы убедиться, что артефакты удалены.**

In [10]:
data['children'].unique()

array([1, 0, 3, 2, 4, 5])

### Удаление пропусков (продолжение)

**Задание 11. Заполните пропуски в столбце `days_employed` медианными значениями по каждого типа занятости `income_type`.**

In [11]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['days_employed'].isna()), 'days_employed'] = \
    data.loc[(data['income_type'] == t), 'days_employed'].median()

**Задание 12. Убедитесь, что все пропуски заполнены. Проверьте себя и ещё раз выведите количество пропущенных значений для каждого столбца с помощью двух методов.**

In [12]:
data.isna().sum()

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

### Изменение типов данных

**Задание 13. Замените вещественный тип данных в столбце `total_income` на целочисленный с помощью метода `astype()`.**

In [13]:
data['total_income'] = data['total_income'].astype(int)

### Обработка дубликатов

**Задание 14. Обработайте неявные дубликаты в столбце `education`. В этом столбце есть одни и те же значения, но записанные по-разному: с использованием заглавных и строчных букв. Приведите их к нижнему регистру.**

In [14]:
data['education'] = data['education'].str.lower()

**Задание 15. Выведите на экран количество строк-дубликатов в данных. Если такие строки присутствуют, удалите их.**

In [15]:
data.duplicated().sum()

71

In [16]:
data = data.drop_duplicates()

### Категоризация данных

**Задание 16. На основании диапазонов, указанных ниже, создайте в датафрейме `data` столбец `total_income_category` с категориями:**

- 0–30000 — `'E'`;
- 30001–50000 — `'D'`;
- 50001–200000 — `'C'`;
- 200001–1000000 — `'B'`;
- 1000001 и выше — `'A'`.


**Например, кредитополучателю с доходом 25000 нужно назначить категорию `'E'`, а клиенту, получающему 235000, — `'B'`. Используйте собственную функцию с именем `categorize_income()` и метод `apply()`.**

In [17]:
def categorize_income(income):
    try:
        if 0 <= income <= 30000:
            return 'E'
        elif 30001 <= income <= 50000:
            return 'D'
        elif 50001 <= income <= 200000:
            return 'C'
        elif 200001 <= income <= 1000000:
            return 'B'
        elif income >= 1000001:
            return 'A'
    except:
        pass

In [18]:
data['total_income_category'] = data['total_income'].apply(categorize_income)

**Задание 17. Выведите на экран перечень уникальных целей взятия кредита из столбца `purpose`.**

In [19]:
data['purpose'].unique()

array(['покупка жилья', 'приобретение автомобиля',
       'дополнительное образование', 'сыграть свадьбу',
       'операции с жильем', 'образование', 'на проведение свадьбы',
       'покупка жилья для семьи', 'покупка недвижимости',
       'покупка коммерческой недвижимости', 'покупка жилой недвижимости',
       'строительство собственной недвижимости', 'недвижимость',
       'строительство недвижимости', 'на покупку подержанного автомобиля',
       'на покупку своего автомобиля',
       'операции с коммерческой недвижимостью',
       'строительство жилой недвижимости', 'жилье',
       'операции со своей недвижимостью', 'автомобили',
       'заняться образованием', 'сделка с подержанным автомобилем',
       'получение образования', 'автомобиль', 'свадьба',
       'получение дополнительного образования', 'покупка своего жилья',
       'операции с недвижимостью', 'получение высшего образования',
       'свой автомобиль', 'сделка с автомобилем',
       'профильное образование', 'высшее об

**Задание 18. Создайте функцию, которая на основании данных из столбца `purpose` сформирует новый столбец `purpose_category`, в который войдут следующие категории:**

- `'операции с автомобилем'`,
- `'операции с недвижимостью'`,
- `'проведение свадьбы'`,
- `'получение образования'`.

**Например, если в столбце `purpose` находится подстрока `'на покупку автомобиля'`, то в столбце `purpose_category` должна появиться строка `'операции с автомобилем'`.**

**Используйте собственную функцию с именем `categorize_purpose()` и метод `apply()`. Изучите данные в столбце `purpose` и определите, какие подстроки помогут вам правильно определить категорию.**

In [20]:
def categorize_purpose(row):
    try:
        if 'автом' in row:
            return 'операции с автомобилем'
        elif 'жил' in row or 'недвиж' in row:
            return 'операции с недвижимостью'
        elif 'свад' in row:
            return 'проведение свадьбы'
        elif 'образов' in row:
            return 'получение образования'
    except:
        return 'нет категории'

In [21]:
data['purpose_category'] = data['purpose'].apply(categorize_purpose)

### Шаг 3. Исследуйте данные и ответьте на вопросы

#### 3.1 Есть ли зависимость между количеством детей и возвратом кредита в срок?

In [22]:
# Ваш код будет здесь. Вы можете создавать новые ячейки.
# группируем таблицу по количеству детей в семье и количеству клиентов банка с учетом возврата кредита в срок 
data_grouped_age = data.groupby(['children']).agg({'family_status_id': 'count', 'debt':'sum',})
# Доля просрочек в каждой группе
data_grouped_age['share_not_ret'] = data_grouped_age['debt'] / data_grouped_age['family_status_id']*100


In [23]:
# выводим таблицу сортируем по убыванию не возвращенных кредитов 
data_grouped_age.sort_values(by='debt', ascending=False)

family_status_id  debt  share_not_ret
children                                       
0                    14091  1063       7.543822
1                     4808   444       9.234609
2                     2052   194       9.454191
3                      330    27       8.181818
4                       41     4       9.756098
5                        9     0       0.000000

**Вывод:** Была выявлена зависимость между количеством детей у клиента и возвратом банковского кредита в срок.
Данные таблицы говорят о том, что клиенты банка взявшие кредит и не имеющие детей, чаще исполняют свои кредитные обязательства перед банком в срок, чем клиенты банка взявшие кредиты и имеющие детей.


Клиенты банка с одним или двумя детьми в семье не возвращают кредиты банку вовремя примерно с одинаковым процентом. Семья с одним ребенком это вторая крупная группа клиентов банка, в нее входит 4808 клиентов. Семья с двумя детьми третья группа по числености клиентов банка, в нее входит 2052 клиента.

Клиенты банка в семье которых три ребенка, возвращают кредиты чаще чем семьи с одним ребенком, но 
данных клиенто всего 330. Чем меньше клиентов тем сильнее влиет каждый кредит возвращенный вовремя на долю возвратов на количество клиентов. 

Клиенты банка в семьях которых четыре ребенка, составляют 41 клиент банка. Эти клиенты имеют самый большой процент не возвращенных кредитов банка. При своей очень не большой числености как калиенты у банка, имеют самый большой процент не возвращенных вовремя кредитов.

Клиенты банка с пятью детьми,вернули все кредиты вовремя, но данная группа состоит всего из 9 клиентов .Является самой малочисленной клиентской группой у банка. 

#### 3.2 Есть ли зависимость между семейным положением и возвратом кредита в срок?

In [24]:
# Ваш код будет здесь. Вы можете создавать новые ячейки
# Посмотрели сколько уникальных семейных статусов есть в таблице
data['family_status'].unique()

array(['женат / замужем', 'гражданский брак', 'вдовец / вдова',
       'в разводе', 'Не женат / не замужем'], dtype=object)

In [25]:
# Сделали все буквы в названии семейных статусов строчными  
data['family_status'] = data['family_status'].str.lower() 

In [26]:
# Нашли зависимость между семейным положением и количеством клиентов в этих строках  и вовзратом кредита в срок
data_grouped_fam_stat = data.groupby(['family_status']).agg({'family_status':'count',
                                                                  'debt':'sum'})

In [27]:
# Доля просрочек в каждой группе
data_grouped_fam_stat['share_not_ret'] = data_grouped_fam_stat['debt'] / data_grouped_fam_stat['family_status']*100

In [28]:
# Вывели результат отсортировав его по убыванию 
data_grouped_fam_stat.sort_values(by='debt',ascending=False)

family_status  debt  share_not_ret
family_status                                            
женат / замужем                12261   927       7.560558
гражданский брак                4134   385       9.313014
не женат / не замужем           2796   273       9.763948
в разводе                       1189    84       7.064760
вдовец / вдова                   951    63       6.624606

**Вывод:** 

Была выявлена зависимость между семейным положеним клиентов и их возвратом кредита в срок.
 
Группа клиентов банка с семейным статусом "женат/замужем" является самой большой, что состовляет 12261 человек. У данной группы клиентов бюанка процент невозвращенных кредитов вовремя составляет 7.56%.

Группа клиентов банка с семейным статусом "гражданский брак" является второй по количеству взявших кредит, и состовляет 4134 человека. У данной группы клиентов процент кредитов, невозвращенных в срок, больше чем у клиентов банка с семейным статусом "женат/замужем", и состовляет 9.3%.

Группа клиентов банка с семейным статусом "не женат/ не замужем", является третьей по количеству клиентов банка, и состовляет 2796 человек. Данная группа имееет самый большой процент невозвращенных кредитов вовремя, что состовляет 9.76% 

Группа клиентов банка с семейным статусом "в разводе", состоящая из 1189 клиентов, не возвращает кредиты вовремя, что составляет 7.06%.

Группа клиентов банка с семейным статусом "вдовец/вдова" является самой малочисленной из всех представленных, и состоит из 951 клиента. Данная группа клиентов имеет самый низкий процент невозвращенных кредитов в срок, который состовляет 6.62%.

#### 3.3 Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

In [29]:
# Ваш код будет здесь. Вы можете создавать новые ячейки
# Нашли зависимость между уровне дохода и возратом кредита в срок 
# Для этого использовали ранее созданный столбец total_income_category,
# который распределил доходы на группы(A,B,C,D,E)
data_grouped_total_income = data.groupby(['total_income_category']).agg({'family_status_id': 'count',
                                                                         'total_income':'median',
                                                                         'debt':'sum'})

In [30]:
# Доля просрочек в каждой группе
data_grouped_total_income['share_not_ret'] = data_grouped_total_income['debt'] / data_grouped_total_income['total_income']*100

In [31]:
# Выводим сгрупированную таблицу и сортируем по убыванию общего медианного дохода по категориям
data_grouped_total_income.sort_values(by='total_income',ascending=False)

family_status_id  total_income  debt  share_not_ret
total_income_category                                                     
A                                    25       1223042     2       0.000164
B                                  5014        257558   354       0.137445
C                                 15921        130355  1353       1.037935
D                                   349         43929    21       0.047804
E                                    22         26431     2       0.007567

**Вывод:** 
Группа 'С' имеет самое большое количество клиентов банка - 15921 человек со средним доходом 130355 рублей. Процент невозвращенных кредитов среди них составляет 1.03%.

Группа 'B' состоит из 5014 клиентов банка при среднем доходе 257558 рублей. Процент невозвращенных кредитов среди них составляет 0.13%.

Клиентов банка категории 'C' - 15921 человек, это больше, чем клиентов банка 'B'- 5014 человек. Но доход клиентов банка 'B'- 257558 рублей, что больше, чем клиентов 'C', где зарплата 130355 рублей.  Процент невозвращенных кредитов клиентами банка вовремя у клиентов группы'B' на 0.13% ниже, чем у клиентов банка группы 'C' 1.03%.

Группы клиентов 'А', 'D', 'E' не были использованы при анализе. Так как в данных группах не достаточно клиентов банка для анализа.



#### 3.4 Как разные цели кредита влияют на его возврат в срок?

In [32]:
# Ваш код будет здесь. Вы можете создавать новые ячейки.
# Нашли зависимость между целями кредита и его возвратом в срок
# Для этого использовали ранее сформированных столбец purpose_category
# В purpose_category собранны все цели клиентов банка и разделены 4 категории по целям
data_groupedpur = data.groupby(['purpose_category']).agg({'family_status_id': 'count', 'debt':'sum'})

In [33]:
# Доля просрочек в каждой группе
data_groupedpur['share_not_ret'] = data_groupedpur['debt'] / data_groupedpur['family_status_id']*100

In [34]:
# Выводим результа сортируем по убыванию 
data_groupedpur.sort_values(by='debt',ascending=False)

family_status_id  debt  share_not_ret
purpose_category                                               
операции с недвижимостью             10751   780       7.255139
операции с автомобилем                4279   400       9.347978
получение образования                 3988   369       9.252758
проведение свадьбы                    2313   183       7.911803

**Вывод:** 
Было выявлено как разные цели кредита влияют на его возврат в срок.
Клиенты банка, которые взяли кредит на 'операции с недвижимостью', являются самой большой группой клиентов банка - 10751 человек. Данная группа показала низкий процент невозвратов кредитов в срок - 7.25%.

Клиенты банка, которые бралил кредиты на 'операции с автомобилем' состовляют 4279 человек, их процент не возрата кредита - 9.34%. Клиенты банка взявшие кредит на 'получение образования' состовляют 3988 человек, их процент не возрата кредита - 9.25%. В данных группах количесво клиентов банка и процент невозвращенных кредитов в срок схож.


Клиенты банка, которые брали кредиты на 'проведение свадьбы', составляет 2313 человек. Данные клиентыы не возвращают кредиты в срок 7.91%. Данный показатель является вторым в сравнении с клиентами банка, которые берут кредит на 'операции с недвижимостью'

#### 3.5 Приведите возможные причины появления пропусков в исходных данных.

*Ответ:* 
Человеческий фактор,
ошибки при выгрузке данных.

#### 3.6 Объясните, почему заполнить пропуски медианным значением — лучшее решение для количественных переменных.

*Ответ:*
Заполнение медианным значением пропусков количественных переменных - лучшее решение, потому что так эти пропусти не могу влият на результат анализа данных и можно делать выводы, в которых нет аномальных значений.  

### Шаг 4: общий вывод.

Напишите ваш общий вывод.
Провели анализ данных клиентов банка по критериям
- Количеством детей и возвратом кредита в срок
- Семейные положение и возвратом кредита в срок
- Уровнем дохода и возвратом кредита в срок
- Цели кредита и влияние на его возврат в срок


- Количеством детей и возвратом кредита в срок
Была выявлена зависимость между количеством детей у клиента и возвратом банковского кредита в срок.
Больше всего берут кредиты клиенты банка, которые не имеют детей, у данных клиентов самая низкий процент не возврата кредита в срок. Реже берут кредиты клиенты банка с 1-м, 2-мя, 4-мя детьми, но процент не возврата достаочной высокий более 9%.

- Семейные положение и возвратом кредита в срок
Была выявлена зависимость между семейный положением и возвратом кредита в срок. 
Клиенты банка, которые имеют семейное положение "женат/замужем", "в разводе", "вдовец / вдова", показывают низкие проценты по невовзрату кредитов в срок, не более 8%.
Клиенты банка,которые имеют семейное положение "гражданский брак" или "не женат / не замужем" показывают проценты по невозврату кредита в срок более 9%.

- Уровнем дохода и возвратом кредита в срок
Была выявлена зависимость между доходом и возвратом кредита в срок
Выявили две группу и зависимость между средним доходом клиента и возвратом клиентом кредита в срок
Были рассмотрены две группы клиентов банка, поделенные по среднему доходу, на 'B' и 'С'.

Процент не возврата кредита в срок у группы клиентов банка 'B', где 5014 человек со средний доходом 257558 рублей, состовляет 0.13%.

Процеент не возврата кредиа в срок у группы клиентов банка 'C', где 15921 человек со средним доходом 
130355 рублей, состаляет 1%. 

В группе 'B' в три раза меньше количество клиентов банка, чем в группе 'C', и доход группы 'B' выше в два раза чем в группе 'C'. Группа 'B' выплачивает кредиты в срок, чаще чем группа 'C'.



- Цели кредита и влияние на его возврат в срок
Была выявлена зависимость между целью кредита и невозврата кредита в срок.

Клиенты банка,которые берут кредиты на 'операции с недвижимостью', состовляет 10751 человек  и имею самую низкую просрочку по кредитам не выплаченным в срок, которая составляет 7.25%.

Клиенов банка, которые берут кредиты 'операции с автомобилем' и 'получение образования', состовляет 4279 и 3988 человек, но данные клиенты банка, имеют самый высокий процент не выплаты кредитов в срок, который состовляет 9.34% и 9.25%.

Клиенты банка, которые берут кредиты на 'проведение свадьбы', состовляет 2313 человек, данная группа является малочисленной из представленных выше, но выплаты по кредитам не в срок составляют 7.91%.


